### PR TITLE
Emscripten build of jsdrafter.js

### PR DESCRIPTION
**Pull request by w-vi**
- See README for build details
- Works nice in node/io.js too
- Size in release build is ~1.1 MB
- Uglify-js to squeeze the size to ~955K
- Circle CI test

Circle CI test needs emsdk.patch and may take up to 2 hrs if all
dependencies need to be built from scratch. Caching is in place though.

### CODE REVIEW

**Comment by w-vi** [comment]
@zdne PING!

**Comment by zdne** [comment]
@w-vi yup, I keep it in mind. It is just the meetup war we want through. Next sprint – starting next week we will get to it. Also what is up with the failed tests?

**Comment by w-vi** [comment]
@zdne : failing test is actually good news my upstream feature request to emscripten sdk was pushed to master so the patch I've made for it can be now removed, hurray. I'll fix it.

**Comment by w-vi** [comment]
@zdne So the failing test is caused now by not being able to git clone the emscripten sources, I have no idea why as locally it works.

**Comment by w-vi** [comment]
what I'm going to try tomorrow is to utilize docker so the emscripten will be in docker image, this might solve a lot of troubles with the ci now as well as in the future.

**Comment by zdne** [comment]
@w-vi :+1:

**Comment by w-vi** [comment]
@zdne Dockerized emscripten in place .... I hope that now we will have much less troubles with it. You can use it as well for testing and building locally, see circle.yml for example usage. Or ask if anything is unclear.

**Comment by w-vi** [comment]
@zdne I' m sorry to ping again but again but I'd like to get it moving as I've got some ideas how to improve it but I don't want to do it unless it has some users to get some initial feedback.

**Comment by zdne** [comment]
@w-vi we still haven't get to it – ETA 2 weeks :(

**Comment by w-vi** [comment]
@zdne OK, so I'll ping again in two weeks :)

File: README.md [hunk 1]
```diff
@@ -93,40 +93,90 @@ Refer to [AST Serialization Media Types](https://github.com/apiaryio/api-bluepri
...
-	```sh
-	$ sudo make install
-	$ drafter --help
-	```
+    ```sh
+    $ sudo make install
+    $ drafter --help
+    ```
+
+## jsdrafter.js
```
Comment (danielgtaylor) [inline README.md]: Is there a reason to prefer `jsdrafter.js` over `drafter.js`?

File: README.md [hunk 2]
```diff
@@ -93,40 +93,90 @@ Refer to [AST Serialization Media Types](https://github.com/apiaryio/api-bluepri
...
+exposes 2 functions `string parseAPIBlueprint(string)` and `int
+validateAPIBlueprint(string)` which serve for getting a parser output
+in JSON format and to just validate the API Blueprint respectively.
+It is build from the C++ source using
+[emscripten](http://kripken.github.io/emscripten-site/).
+
+
+### Build jsdrafter.js
+
+*Unfortunately it works only on *nix environment at the moment.*
```
Comment (danielgtaylor) [inline README.md]: Clarify that it is only **building** that requires *nix. Using the library works everywhere.

File: circle.yml [hunk 3]
```diff
@@ -1,12 +1,34 @@
+general:
+  artifacts:
+  - ~/drafter/emcc/jsdrafter.js
+  - ~/drafter/emcc/jsdrafter.js.mem
```
Comment (danielgtaylor) [inline circle.yml]: Is there somewhere to download these? Maybe we should link to them in the README?

File: circle_emsdk.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+mkdir -p emsdk
+tar -xzf emsdk-portable.tar.gz -C emsdk --strip-components=1
+cd emsdk
+./emsdk update
+sed -i "s|fastcomp/|fastcomp|g" emsdk_manifest.json
+sed -i "s|clang/|clang|g" emsdk_manifest.json
+./emsdk install latest -j1
+./emsdk activate latest
+cp ~/.emscripten .
+cd ../
```
Comment (danielgtaylor) [inline circle_emsdk.sh]: Is this file still required with Docker?

File: emcc/post.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    buffer = _malloc(text.length+1);
+    writeStringToMemory(text, buffer);
+    res = _drafter_c_parse(buffer, 0, 0);
+    _free(buffer);
+    return res ;
+};
+
+var parseAPIBlueprint = Module['parseAPIBlueprint'];
+var validateAPIBlueprint = Module['validateAPIBlueprint'];
```
Comment (danielgtaylor) [inline emcc/post.js]: Can we update these to mimic the API used by Drafter and Protagonist? For example:

```js
var drafter = require('drafter.js');

var result;
var options = {
  exportSourcemap: true,
  type: 'ast'
}

try {
  result = drafter.parse('# My Blueprint', options);
} catch (err) {
  // ...
}
```

File: emcc/pre.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+var Module = {
+    'ready': false,
+    'print': function(text) { console.log('jsdrafter.js: ' + text); },
+    'printErr': function(text) { console.error('jsdrafter.js: ' + text); },
```
Comment (danielgtaylor) [inline emcc/pre.js]: Might consider renaming these to `log` and `logErr`. Presumably when you override them you want to log information and not necessarily print it to the screen.

**Comment by danielgtaylor** [comment]
@w-vi this looks great and I'd love to see it get merged in soon so we can move forward with a JS-only option for parsing. My main concern is the public interface - I'd much rather see it mimic what we have today in the drafter and protagonist modules. How hard is it to update that?

File: README.md (hunk 1 above)
Comment (w-vi) [inline README.md]: There was a `drafter.js` in use at the time when I started doing it, I'd prefer to call it `drafter.js` too so happy to rename it.

File: circle.yml (hunk 3 above)
Comment (w-vi) [inline circle.yml]: No there is no place to download them, but can be easily done and I suppose that any `drafter` release will have a `drafter.js` company.

File: circle_emsdk.sh (hunk 4 above)
Comment (w-vi) [inline circle_emsdk.sh]: Probably not, I'll have a look and remove it if not needed.

**Comment by w-vi** [comment]
@danielgtaylor Updating the interface is fairly easy if the C API provides the same information. I'll have a look.
I'm really excited that finally it is getting some attention again. Thanks for the comments, I'll address them as soon as possible.

**Comment by kylef** [comment]
Closing since this has been superseded by #144. :balloon:
